### PR TITLE
Enhance forward index reader for better performance

### PR DESCRIPTION
**Pull request by Jackie-Jiang**
## Description
Enhance the forward index reader by specializing the case for each numBits (from 1 to 31) compression.

The int value reader has 3 methods:
- Bulk read 32 values
- Read value without buffer boundary check
- Read value with buffer boundary check

The new forward index reader uses the 3 methods based on the passed in doc ids:
- For sequential doc ids, use bulk read
- For non-sequential doc ids, use unchecked read if possible, and only use checked read for the last few values

We observe up to 7x improvement for the reading, detailed perf number can be found here: https://docs.google.com/spreadsheets/d/1o66G7Jw_7K4swEcp1FYCzgr0MF9JBAAM6KMGOXBSdIA/edit?usp=sharing

### CODE REVIEW

**Review by mayankshriv — COMMENTED** [review]
Seems like we are moving to `FixedBitSVForwardIndexReaderV2` by default. Just in case there's a production issue, the only option would be to rollback?

File: pinot-core/src/test/java/org/apache/pinot/core/segment/index/readers/forward/FixedBitSVForwardIndexReaderV2Test.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+import static org.testng.Assert.assertEquals;
+
+
+public class FixedBitSVForwardIndexReaderV2Test {
+  private static final File INDEX_DIR = new File(FileUtils.getTempDirectory(), "FixedBitIntReaderTest");
+  private static final int NUM_VALUES = 99999;
+  private static final int NUM_DOC_IDS = 10000;
```
Comment (mayankshriv) [inline pinot-core/src/test/java/org/apache/pinot/core/segment/index/readers/forward/FixedBitSVForwardIndexReaderV2Test.java:37]: Nit suggestion: Use prime numbers?

File: pinot-core/src/test/java/org/apache/pinot/core/segment/index/readers/forward/FixedBitSVForwardIndexReaderV2Test.java [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+  private final int[] _sparseDocIds = new int[NUM_DOC_IDS];
+  private final int[] _lastSequentialDocIds = new int[NUM_DOC_IDS];
+  private final int[] _dictIdBuffer = new int[NUM_DOC_IDS];
+
+  @BeforeClass
+  public void setUp()
+      throws IOException {
+    FileUtils.forceMkdir(INDEX_DIR);
+
+    int sequentialDocId = RANDOM.nextInt(32);
```
Comment (mayankshriv) [inline pinot-core/src/test/java/org/apache/pinot/core/segment/index/readers/forward/FixedBitSVForwardIndexReaderV2Test.java:50]: Would it be better to use explicit values instead of random?
For example, we know all the corner cases would be around bits being split across read boundaries. We can then programmatically generate all the splits.
We could have random as well as explict.

File: pinot-core/src/test/java/org/apache/pinot/core/segment/index/readers/forward/FixedBitSVForwardIndexReaderV2Test.java (hunk 1 above)
Comment (Jackie-Jiang) [reply pinot-core/src/test/java/org/apache/pinot/core/segment/index/readers/forward/FixedBitSVForwardIndexReaderV2Test.java:37]: This is the `DocIdSetPlanNode.MAX_DOC_PER_CALL`, replaced it with the constant

File: pinot-core/src/test/java/org/apache/pinot/core/segment/index/readers/forward/FixedBitSVForwardIndexReaderV2Test.java (hunk 2 above)
Comment (Jackie-Jiang) [reply pinot-core/src/test/java/org/apache/pinot/core/segment/index/readers/forward/FixedBitSVForwardIndexReaderV2Test.java:50]: That's why I explicitly tested reading the last doc ids (`_lastSequentialDocIds`).
Modified the test to test sequentialDocId starting from 0 - 31

**Comment by Jackie-Jiang** [comment]
> Seems like we are moving to `FixedBitSVForwardIndexReaderV2` by default. Just in case there's a production issue, the only option would be to rollback?

@mayankshriv Yes. I don't think it's worth adding a config to switch between readers. Per the perf, the new reader has way better performance in all scenarios.

**Review by kishoreg — COMMENTED** [review]
where is the logic to decide sparse vs sequential vs dense read

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/loader/invertedindex/TextIndexHandler.java [hunk 3]
```diff
@@ -51,7 +51,7 @@
 import org.apache.pinot.core.segment.index.readers.ForwardIndexReaderContext;
 import org.apache.pinot.core.segment.index.readers.StringDictionary;
 import org.apache.pinot.core.segment.index.readers.forward.BaseChunkSVForwardIndexReader.ChunkReaderContext;
-import org.apache.pinot.core.segment.index.readers.forward.FixedBitSVForwardIndexReader;
+import org.apache.pinot.core.segment.index.readers.forward.FixedBitSVForwardIndexReaderV2;
```
[the remark this answers is not in the document]
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/segment/index/loader/invertedindex/TextIndexHandler.java:54]: V1 is replaced with V2. Will do the clean up as a separate PR

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/forward/FixedBitSVForwardIndexReaderV2.java [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+  @Override
+  public void readDictIds(int[] docIds, int length, int[] dictIdBuffer, ForwardIndexReaderContext context) {
+    int firstDocId = docIds[0];
+    int lastDocId = docIds[length - 1];
+    int index = 0;
+
+    // Use bulk read if the doc ids are sequential
+    if (lastDocId - firstDocId + 1 == length && length >= 64) {
+      int[] buffer = new int[32];
```
[the remark this answers is not in the document]
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/forward/FixedBitSVForwardIndexReaderV2.java:69]: This is a very small array and the overhead is negligible. We create the buffer per block (10000 docs) only when bulk read is needed

**Comment by Jackie-Jiang** [comment]
> where is the logic to decide sparse vs sequential vs dense read

In `FixedBitSVForwardIndexReaderV2`, we use bulk read only for sequential doc ids. Per the benchmark, using bulk read for dense doc ids performs worse than unchecked read, so there is no difference for reading sparse and dense doc ids

File: pinot-core/src/test/java/org/apache/pinot/core/io/reader/impl/FixedBitIntReaderTest.java [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+import static org.testng.Assert.assertEquals;
+
+
+public class FixedBitIntReaderTest {
+  private static final File INDEX_DIR = new File(FileUtils.getTempDirectory(), "FixedBitIntReaderTest");
+  private static final int NUM_VALUES = 95;
+  private static final Random RANDOM = new Random();
```
Comment (siddharthteotia) [inline pinot-core/src/test/java/org/apache/pinot/core/io/reader/impl/FixedBitIntReaderTest.java:37]: How is this covering all the numbits cases for new APIs?

Comment (siddharthteotia) [reply pinot-core/src/test/java/org/apache/pinot/core/io/reader/impl/FixedBitIntReaderTest.java:37]: We should add tests for new APIs for all numbits right?

File: pinot-core/src/test/java/org/apache/pinot/core/segment/index/readers/forward/FixedBitSVForwardIndexReaderV2Test.java [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+import static org.testng.Assert.assertEquals;
+
+
+public class FixedBitSVForwardIndexReaderV2Test {
+  private static final File INDEX_DIR = new File(FileUtils.getTempDirectory(), "FixedBitIntReaderTest");
+  private static final int NUM_VALUES = 99_999;
```
Comment (siddharthteotia) [inline pinot-core/src/test/java/org/apache/pinot/core/segment/index/readers/forward/FixedBitSVForwardIndexReaderV2Test.java:37]: We should add tests for all numBits and all possible docID ranges to ensure the new APIs are correctly decoding (and there is no possibility of segfault)

**Comment by siddharthteotia** [comment]
I think the JMH perf is probably a reasonable indicator of the benefit for single read APIs. However,  I don't think we have added exhaustive unit tests for all numBits and different docId ranges covering all cases to ensure these are all functionally correct and won't segfault.

```
@Override
    public int read(int index) {
      int offset = index >>> 3;
      int bitOffsetInByte = index & 0x7;
      return (_dataBuffer.getByte(offset) >>> (7 - bitOffsetInByte)) & 0x1;
    }

    @Override
    public int readUnchecked(int index) {
      int offset = index >>> 3;
      int bitOffsetInByte = index & 0x7;
      return (_dataBuffer.getByte(offset) >>> (7 - bitOffsetInByte)) & 0x1;
    }

```

For the getDictIds function:

```
@Override
  public void readDictIds(int[] docIds, int length, int[] dictIdBuffer, ForwardIndexReaderContext context) 
```

- Is it correct to say that new bulk read32()  will be used here only if the docIDs array has sequential (consecutive) ids. Otherwise, we will use the new single read APIs?
- The readUnchecked() has the potential to segfault / core dump?
- There is a chance that perf wise, we might see -ve impact due to memcpy (especially for high throughput use cases). So for this API at least, I don't think JMH is a reasonable indicator of better performance always. It is worthwhile to do validation on a prod-like workload.

File: pinot-core/src/test/java/org/apache/pinot/core/io/reader/impl/FixedBitIntReaderTest.java (hunk 5 above)
Comment (Jackie-Jiang) [reply pinot-core/src/test/java/org/apache/pinot/core/io/reader/impl/FixedBitIntReaderTest.java:37]: This is just the random value generator. All the bits are tested. Check the for look on line 55

File: pinot-core/src/test/java/org/apache/pinot/core/segment/index/readers/forward/FixedBitSVForwardIndexReaderV2Test.java (hunk 6 above)
Comment (Jackie-Jiang) [reply pinot-core/src/test/java/org/apache/pinot/core/segment/index/readers/forward/FixedBitSVForwardIndexReaderV2Test.java:37]: It tests all the bits (for loop on line 77), and sequential reads starting at index 0-31, and also the last 10000 docs. IMO it is quite thorough.

**Comment by Jackie-Jiang** [comment]
@siddharthteotia Please check my reply under the comments. I think the tests are quite thorough, and covers all the bits and different start indices.

For your questions:

> Is it correct to say that new bulk read32() will be used here only if the docIDs array has sequential (consecutive) ids. Otherwise, we will use the new single read APIs?

Yes

> The readUnchecked() has the potential to segfault / core dump?

If you use readUnchecked() for the last 2 values, it can read over the boundary (explained in the javadoc). This is handled within the `FixedBitSVForwardIndexReaderV2`

> There is a chance that perf wise, we might see -ve impact due to memcpy (especially for high throughput use cases). So for this API at least, I don't think JMH is a reasonable indicator of better performance always. It is worthwhile to do validation on a prod-like workload.

Why is memcpy involved here? The new reader won't read extra values because the bulk read is only used for sequential ids. I already asked @pradeepgv42 to try this, and he confirmed up to 45% query latency drop comparing with the current reader.

**Comment by siddharthteotia** [comment]
> @siddharthteotia Please check my reply under the comments. I think the tests are quite thorough, and covers all the bits and different start indices.
> 
> For your questions:
> 
> > Is it correct to say that new bulk read32() will be used here only if the docIDs array has sequential (consecutive) ids. Otherwise, we will use the new single read APIs?
> 
> Yes
> 
> > The readUnchecked() has the potential to segfault / core dump?
> 
> If you use readUnchecked() for the last 2 values, it can read over the boundary (explained in the javadoc). This is handled within the `FixedBitSVForwardIndexReaderV2`
> 
> > There is a chance that perf wise, we might see -ve impact due to memcpy (especially for high throughput use cases). So for this API at least, I don't think JMH is a reasonable indicator of better performance always. It is worthwhile to do validation on a prod-like workload.
> 
> Why is memcpy involved here? The new reader won't read extra values because the bulk read is only used for sequential ids. I already asked @pradeepgv42 to try this, and he confirmed up to 45% query latency drop comparing with the current reader.

This is a memcpy?

`for (int i = bulkStartIndex; i < bulkEndIndex; i += 32) {
        _reader.read32(i, buffer);
        System.arraycopy(buffer, 0, dictIdBuffer, index, 32);
        index += 32;
      }`

Just to be on the same page, bulkRead will be used only for queries like select * without filter or any query without filter or any query with range filter on sorted column?

**Comment by Jackie-Jiang** [comment]
> > @siddharthteotia Please check my reply under the comments. I think the tests are quite thorough, and covers all the bits and different start indices.
> > For your questions:
> > > Is it correct to say that new bulk read32() will be used here only if the docIDs array has sequential (consecutive) ids. Otherwise, we will use the new single read APIs?
> > 
> > 
> > Yes
> > > The readUnchecked() has the potential to segfault / core dump?
> > 
> > 
> > If you use readUnchecked() for the last 2 values, it can read over the boundary (explained in the javadoc). This is handled within the `FixedBitSVForwardIndexReaderV2`
> > > There is a chance that perf wise, we might see -ve impact due to memcpy (especially for high throughput use cases). So for this API at least, I don't think JMH is a reasonable indicator of better performance always. It is worthwhile to do validation on a prod-like workload.
> > 
> > 
> > Why is memcpy involved here? The new reader won't read extra values because the bulk read is only used for sequential ids. I already asked @pradeepgv42 to try this, and he confirmed up to 45% query latency drop comparing with the current reader.
> 
> This is a memcpy
> 
> ```
> `for (int i = bulkStartIndex; i < bulkEndIndex; i += 32) {
>         _reader.read32(i, buffer);
>         System.arraycopy(buffer, 0, dictIdBuffer, index, 32);
>         index += 32;
>       }`
> ```
> 
> Just to be on the same page, bulkRead will be used only for queries like select * without filter or any query without filter or any query with range filter on sorted column?

Yes, correct

**Review by siddharthteotia — APPROVED** [review]
LGTM. 

Thanks for working on this. 

It will be great if we can the following PR before this one to help cut a internal release 

https://github.com/apache/incubator-pinot/pull/6284
